docs/source/step-by-step.rst

```diff
@@ -1,15 +1,760 @@
+
+^^^^^^^^^^^^^^^^^^^^^^
+nMigen CFU Development
+^^^^^^^^^^^^^^^^^^^^^^
+
+There is a fairly robust framework for building a CFU in nMigen. Inside of
+``<CFU-Playground root>/python/nmigen_cfu`` there are a set of helper files
+that you can ``import`` in your code. It's best to read through the doc
+comments in ``<CFU-Playground root>/python/nmigen_cfu/cfu.py`` and
+``<CFU-Playground root>/python/nmigen_cfu/util.py`` before starting
+development, but you should be able to get a reasonable understanding of the
+framework through this example.
+
+.. code-block:: python
+
+    from nmigen import C, Module, Signal, signed
+    from nmigen_cfu import all_words, Cfu, InstructionBase, InstructionTestBase, pack_vals
+    import unittest
+
+
+    # Custom instruction inherits from the InstructionBase class.
+    class SimdMac(InstructionBase):
+        def __init__(self, input_offset=128) -> None:
+            super().__init__()
+
+            self.input_offset = C(input_offset, signed(9))
+
+        # `elab` method implements the logic of the instruction.
+        def elab(self, m: Module) -> None:
+            words = lambda s: all_words(s, 8)
+
+            # SIMD multiply step:
+            self.prods = [Signal(signed(16)) for _ in range(4)]
+            for prod, w0, w1 in zip(self.prods, words(self.in0), words(self.in1)):
+                m.d.comb += prod.eq(
+                    (w0.as_signed() + self.input_offset) * w1.as_signed())
+
+            m.d.sync += self.done.eq(0)
+            # self.start signal high for one cycle when instruction started.
+            with m.If(self.start):
+                with m.If(self.funct7):
+                    m.d.sync += self.output.eq(0)
+                with m.Else():
+                    # Accumulate step:
+                    m.d.sync += self.output.eq(self.output + sum(self.prods))
+                # self.done signal indicates instruction is completed.
+                m.d.sync += self.done.eq(1)
+
+
+    # Tests for the instruction inherit from InstructionTestBase class.
+    class SimdMacTest(InstructionTestBase):
+        def create_dut(self):
+            return SimdMac()
+
+        def test(self):
+            # self.verify method steps through expected inputs and outputs.
+            self.verify([
+                (1, 0, 0, 0),  # reset
+                (0, pack_vals(-128, 0, 0, 1), pack_vals(111, 0, 0, 1), 129 * 1),
+                (0, pack_vals(0, -128, 1, 0), pack_vals(0, 52, 1, 0), 129 * 2),
+                (0, pack_vals(0, 1, 0, 0), pack_vals(0, 1, 0, 0), 129 * 3),
+                (0, pack_vals(1, 0, 0, 0), pack_vals(1, 0, 0, 0), 129 * 4),
+                (0, pack_vals(0, 0, 0, 0), pack_vals(0, 0, 0, 0), 129 * 4),
+                (0, pack_vals(0, 0, 0, 0), pack_vals(-5, 0, 0, 0), 0xffffff84),
+                (1, 0, 0, 0),  # reset
+                (0, pack_vals(-12, -128, -88, -128), pack_vals(-1, -7, -16,
+                                                            15), 0xfffffd0c),
```

Comment:
It would be good to see one test case each for maximum and minimum values: ie. all input values 127 or -128.